frontend/lib/resolution/copy-elision.cpp

```diff
@@ -428,6 +446,19 @@ void FindElidedCopies::handleConditional(const Conditional* cond, RV& rv) {
   } else if (elseFrame == nullptr && thenReturnsThrows) {
     // then returns, no else block, so then copy elisions
     // have already been saved and nothing else to do
+  } else if (thenFrame != nullptr && thenFrame->paramTrue) {
```

Comment:
Reading through this whole change of if-expressions (which I know predateds your PR) is giving me some serious trouble. I'm almost certain, for instance, that the `param true` case for `elseFrame` is missing. But my goodness do I not want to add another branch here.

I wonder if there's a way to generalize all of this in such a way that also makes the code for `select` (which in some way is a glorified if-then-else) easier to write, too.

Reply:
Okay, I think given a general "pick one of these" list of frames, you can get away with something like this:

```C++
void FindElidedCopies::handleDisjunction(VarFrame* parentFrame, const std::vector<VarFrame*>& frames) {
  // if any frame is 'param true', it's the only one that matters.
  for (auto frame : frames) {
    if (frame->paramTrue) {
      // yielding to allElidedCopyFromIds is already handled during return.
      if (frame->returnsOrThrows) return;
      
      // yield all local variables in the frame
      saveLocalVarElidedCopies(frame);

      // Promote outer variable access
      // (* code from line 455 *)
      
      return;
    }
  }

  // No 'param true' branch. For branches that return, we have already yielded / promoted all copy points, so we don't have to look at them at all.
  std::vector<VarFrame*> nonReturningFrames;
  for (auto frame : frames) {
    if (!frame.returnsOrThrows) nonReturningFrames.push_back(frame);
  }

  // Now, note all variables that are elided in _all_ non-returning branches.
  std::unordered_map<ID, size_t> idCounts;
  for (auto frame : nonReturningFrames) {
    for(auto pair : frame->copyElisionState) {
      if (pair.second.lastIsCopy) idCounts[pair.first] += 1;
    }
  }
  for (auto pair : idCounts) {
    // Variable didn't occur in all frames, it doesn't get promoted.
    if (pair.second != nonReturningFrames.size()) continue;

    // Populate the parent frame
    CopyElisionState& parentState = parentFrame->copyElisionState[pair.first];
    parentState.lastIsCopy = true;
    for (auto frame : nonReturningFrames) {
      CopyElisionState& frameState = frame->copyElisionState.at(pair.first);
      parentState.points.insert(frameState.points.begin(), frameState.points.end());
    }
  }
}
```

Given this, the if/else code and the select code can both be written to just push their frames into a vector, and then call this.

Reply:
So thinking about the addition of the `total` argument for this method and when it is/isn't true. Our quick discussion was "has an else" or "has an otherwise," which isn't quite right. Example:
```
proc test() {
          var c: int = 0;
          var x: int = 0;
          var y: int;
          select 2 {
            when c {
              y=x;
            }
            when 2 {
              y=x;
            }
          }
        }
```
The select statement is total, but it doesn't have an otherwise. The correct condition needs to include "has a paramTrue condition." I lean towards storing that aspect as a field on the frame, but can also just recalculate it where needed. What do you think?

Reply:
You're right -- IMO, the _argument_ 'total' should specifically refer to whether or not there's an `else`/`otherwise` (taking a cue from math -- a function is total if an output exists for every input, not just for current known value). So I'd say, compute the "any branch is param true" check in `handleDisjunction`.